Install Bundler version from `Gemfile.lock` + `add_ruby_env_info`

mokagio opened the pull request:
Despite the `gem install bundler` call in `install_gems` we recently see a come back of the `Gem::LoadError: You have already activated bundler...` error from https://github.com/Automattic/bash-cache-buildkite-plugin/issues/16 .

~~I don't know how to fix the issue, but this PR adds info dumps that will hopefully help understand more about its cause.~~

Update: The current best solution is to install the same bundler version as the one tracked in `Gemfile.lock` (original suggestion from #16). The approach has been implemented here and verified [in this Woo demo PR](https://github.com/woocommerce/woocommerce-ios/pull/7820) were the installing the Bundler version from `Gemfile.lock` didn't result in the `Gem::LoadError...` error. See build [here](https://buildkite.com/automattic/woocommerce-ios/builds/8248#0183aeb7-19b5-4ab7-a5be-ec909228ed4c).

<img width="692" alt="image" src="https://user-images.githubusercontent.com/1218433/194400366-9220b2cd-31e4-4e82-9cdd-ff5c3dff634b.png">

These debug calls already revealed something interesting: CI has _two_ `default` Bundler versions, which seems like a probable cause for the "you have already activated..." version issue.

I couldn't find an explanation for that `default` in the `gem list --help`, so I [posted a StackOverflow question](https://stackoverflow.com/questions/73967910/how-can-gem-list-show-multiple-default-versions-for-a-gem) about it.

mokagio committed:
Add new command `dump_ruby_env_info` to help debugging

mokagio committed:
Run `rbenv rehash` after `gem install bundler` + log debug info

mokagio committed:
Add optional prefix to `dump_ruby_env_info` message

This way we can differentiate between multiple calls in the same step

mokagio committed:
Dump Ruby env info at the very beginning of `install_gems`

mokagio committed:
Add a prefix to differentiate info dumps before and after `rbenv rehash`

mokagio committed:
Properly unset `x` in `dump_ruby_env_info`

mokagio committed:
Add `cat .ruby-version` to Ruby info dump

mokagio committed:
Remove `rbenv rehash` call as it didn't make any runtime change

This build shows the info dump being the same before and after the call:
https://buildkite.com/automattic/woocommerce-ios/builds/8225#0183ab0a-a77e-41e6-8338-23ba89e9a032

mokagio commented on bin/dump_ruby_env_info:
```diff
@@ -0,0 +1,23 @@
+
+which bundle
+gem list bundler
+bundle version
```
Feel free to add additional info dump commands if you can think of any 🙇‍♂️

mokagio committed:
Try – Uninstall all Bundler version before installing the latest one

mokagio committed:
Install Bundler form `Gemfile.lock` not the latest

AliSoftware commented on bin/install_gems:
```diff
@@ -1,16 +1,22 @@
-# See https://github.com/Automattic/bash-cache-buildkite-plugin/issues/16  for more details on why this is recommended.
-gem install bundler
+# See https://github.com/Automattic/bash-cache-buildkite-plugin/issues/16 for more details on why this is recommended.
+gem install bundler:"$( (tail -1 | sed 's/ *//g') < Gemfile.lock)"
```
```suggestion
GEMFILE_LOCK_BUNDLER_VERSION=$(sed -n -e '/BUNDLED WITH/{n;s/ *//p;}' Gemfile.lock)
gem install bundler -v "$GEMFILE_LOCK_BUNDLER_VERSION"
```

mokagio replied:
Done in [`c9b9d89` (#32)](https://github.com/Automattic/bash-cache-buildkite-plugin/pull/32/commits/c9b9d89de57896e7567a58a666a48a090d6c689f)

mokagio committed:
Update Bundler version parsing to specifically look for `BUNDLED_WITH`

mokagio committed:
Remove a Ruby info dump that didn't add much value

None of the operations in `restore_cache` change the Ruby environment

mokagio committed:
Add note regarding `gem install bundler --version` error handling

mokagio commented on bin/install_gems:
```diff
@@ -1,16 +1,23 @@
+# See https://github.com/Automattic/bash-cache-buildkite-plugin/issues/16 for more details.
+GEMFILE_LOCK_BUNDLER_VERSION=$(sed -n -e '/BUNDLED WITH/{n;s/ *//p;}' Gemfile.lock)
+# The command will fail if the `--version` parameter is empty; we can omit checking for that ourselves.
+gem install bundler --version "$GEMFILE_LOCK_BUNDLER_VERSION"
```
@crazytonyli 's work in 114-gh-Automattic/buildkite-ci should address the issue this change works around at the root, but I think it's still useful to respect the Bundler version set in the Gemfile.lock.

AliSoftware commented on bin/install_gems:
```diff
@@ -1,16 +1,23 @@
 #!/bin/bash -eu
 
+dump_ruby_env_info "before running $(basename "$0")"
+
```
I wonder if we still want to keep those debug dumps in the final `install_gems` action there?

Personally I'd remove both calls (before + after the `gem install bundler --version …`) to reduce verbosity — one can always temporarily add a call to `dump_ruby_env_info` in their pipeline directly if they need to debug Ruby issues on a given build later.

And if you still think it's useful to keep the dump of the whole Ruby env on every builds and steps calling `install_gem`, I'd still remove that initial one to keep only the one after we did the `gem install`

mokagio replied:
My rationale for having them both is to see the before-after state. It was useful to me when trying to understand the issue, to see which was the selected Bundler as the pipeline started.

But given the fix in [114-gh-Automattic/buildkite-ci](https://href.li/?https://github.com/Automattic/buildkite-ci/issues/114) it's probably safe to remove the info logs, even though I don't think they made the logs particularly more verbose, thanks to Buildkite's logs grouping syntax–although I guess one could argue they _would_ make the raw logs more verbose.

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/more-bundler-info

mokagio committed:
Remove Ruby info dumps from `install_gems`

They were there to debug an issue that has been fixed elsewhere. We can
always add one or both of them back if we feel like we need more info in
the logs.
